add async controller design draft (#48)
ros-controls/roadmap: design_drafts/async_controller.md

The only method of the wrapped controller that we’ll run in a separate thread, is update().
The rest of the method calls will be forwarded from the AsyncControllerWrapper to the wrapped controller in the main thread.
We must make sure that the asynchronous update() is not being executed when forwarding other methods.

### StateInterfaces and CommandInterfaces
The second risk we want to control, is that our asynchronous controller has access to State and Command Interfaces, and uses them from its own thread, causing for instance concurrent write of a CommandInterface while the Actuator is reading it.

To avoid this, we can create intermediate Interfaces, stored in the AsyncControllerWrapper, that act as a bridge between the wrapped controller and the real State and Command Interfaces.
This bridge should allow thread safe data storage, and non blocking operations on the ros_control thread. Some containers for this purpose exist in: https://github.com/ros-controls/realtime_tools/tree/ros2_devel/include/realtime_tools

These interfaces should be created on the assign_interfaces method, and removed on the release_interfaces method. But at the moment those methods are not virtual.

> I’ll assume they will be changed to virtual, but if that cannot be done for ABI compatibility issues, the same could be done in the on_activate and on_deactivate, although it might be less correct.

Reviewer: I don't really count this as an argument. We should do this in a “correct” way. Regardless of ABI compatibility.

Author: had to be listed for correctness ;)

Author: ```suggestion
These interfaces should be created on the assign_interfaces method, and removed on the release_interfaces method. But at the moment those methods are not virtual, that will need to change.
```
